### PR TITLE
KAFKA-6998: Disable Caching when max.cache.bytes are zero.

### PR DESCRIPTION
**Pull request by guozhangwang**
1. As titled, add a `rewriteTopology` that 1) sets application id, 2) maybe disable caching, 3) adjust for source KTable. This optimization can hence be applied for both DSL or PAPI generated Topology.

2. Defer the building of globalStateStores in `rewriteTopology` so that we can also disable caching. But we still need to build the state stores before `InternalTopologyBuilder.build()` since we should only build global stores once for all threads.

3. Added withCachingDisabled to StoreBuilder, it is a public API change.

4. [Optional] Fixed unit test config setting functionalities, and set the necessary config to shorten the unit test latency (now it reduces from 5min to 3.5min on my laptop).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1752]: This is for fixing https://github.com/apache/kafka/pull/5087/files#r209391022. cc @hachikuji

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:640]: Move this logic inside rewriteTopology, ditto below for setApplicationId.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:70]: This is for deferring the build of global stores.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:89]: This is the config to shorten integration test, ditto elsewhere.

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java:49]: This is a minor cleanup for consolidating `getStreamsConfig()` logic across unit tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:363]: Can this be done for trunk ?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:363]: Yes, I think we are going to do that right after #5451 is merged, and then we can move this into the logical plan optimization.

**Comment by guozhangwang** [comment]
Simple benchmark with max.cache.bytes set to 0, comparing this branch:

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-08-10--001.1533946299--guozhangwang--K6998-remove-cache-layer-0cache--39992c1/report.html

with trunk:

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-08-10--001.1533945494--guozhangwang--trunk-0cache--7fc8509/report.html

The results shows that when caching is off, this branch is consistently better than trunk, resulting in 10 - 50 percent performance boost.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:89]: very nice!

**Review by vvcephei — APPROVED** [review]
LGTM. Thanks!

**Comment by tedyu** [comment]
+1

**Review by bbejeck — APPROVED** [review]
Thanks, @guozhangwang  LGTM.  Since https://github.com/apache/kafka/pull/5453 is merged to trunk I think we could do a follow-on PR and we don't have to wait for 5451 to get merged.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: nit: fix indention

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:167]: Why do we want do disable caching for this test now? This seems to be a change in the test itself rather a set setup change?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:256]: as above? (more below -- don't put comment each time)

**Review by mjsax — APPROVED** [review]
Just some clarification questions.

LGTM.

**Comment by mjsax** [comment]
@guozhangwang Feel free to merge after rebasing and green Jenkins build

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: Ack.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:167]: I'm actually not changing the behavior :) The original `getStreamsConfig()` with did lots of hard-coded customization here: https://github.com/apache/kafka/pull/5488/files/f5d72d7579b31fc6bc793767161f0312457a29db#diff-483f89729093ad48aa4b2451decc095eL48

Which I've removed, the motivation is to keep the util functions more intuitive: now we have a few overloads which all calls the underlying

```
Properties getStreamsConfig(final String applicationId,
                                               final String bootstrapServers,
                                               final String keySerdeClassName,
                                              final String valueSerdeClassName,
                                              final Properties additional)
```

which does not make any customization any more. And because of that I moved all the customizations in the caller.

**Comment by guozhangwang** [comment]
Will merge this after KIP-356 gets accepted.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:167]: Ack. Thanks for clarification.

**Pull request merged**
